### ISSUE TITLE
receive op  data had split more times with not local？

### ISSUE THREAD

**Issue by aeroyu** [share/sharedb#396]
i submit op like 
[
  {
    "sd": "checkable-list-item",
    "p": [
      "blocks",
      0,
      "type",
      0
    ]
  },
  {
    "si": "unstyled",
    "p": [
      "blocks",
      0,
      "type",
      0
    ]
  }
]
and local receive op event op is the same with send data，
but with other client receive op，has split two time，[
  {
    "sd": "checkable-list-item",
    "p": [
      "blocks",
      0,
      "type",
      0
    ]
  }] and [{
    "si": "unstyled",
    "p": [
      "blocks",
      0,
      "type",
      0
    ]
  }
]

how can i receive op normal

**Comment by alecgibson** [comment]
Could you please elaborate on your use case? Why is it important to have these delivered as a single object?

**Comment by aeroyu** [comment]
> Could you please elaborate on your use case? Why is it important to have these delivered as a single object?

i use sharedb with draft js，when the op is split and setState，will not work，ex `[
  {
    "sd": "unstyled",
    "p": [
      "blocks",
      0,
      "type",
      0
    ]
  },
  {
    "si": "checkable-list-item",
    "p": [
      "blocks",
      0,
      "type",
      0
    ]
  }
]` 
the first op （delete type unstyled） merge into local ，but it is not my expect，because ，draft js block type can not be empty，and will reset type to unstyled，after apply the second op , the type result is “checkable-list-itemunstyled”

**Comment by aeroyu** [comment]
> Could you please elaborate on your use case? Why is it important to have these delivered as a single object?

i have solve it，i solution is set the uuid and op_size to op model，when receive by other side，i get uuid and op_size to merge the ops  and do one setState，it work！any best  solution？

**Comment by alecgibson** [comment]
Glad to hear you solved it. Honestly off the top of my head I'm not sure there's any (easy) way around it. If you're interested, here's the code that's responsible for that behaviour: https://github.com/share/sharedb/blob/c711cfcb777213d193b1f4a101125e8f6e8e6864/lib/client/doc.js#L584-L620

You'll see it performs this check:

```js
if (!source && this.type === types.defaultType && op.op.length > 1)
```

and then "shatters" the op.

So one potential way around this would be to register second `json0` type, which _isn't_ the default type, and use that, but it's pretty hacky.

You could also try switching to the [`json1`][1] type, which seems to have more semantic operations (including replace).

The final thing — which I _think_ is what you've landed on? — is to buffer incoming ops and batch them together.

Of course you could always raise a pull request to disable this behaviour using some sort of config flag!

[1]: https://github.com/ottypes/json1

**Comment by aeroyu** [comment]
> Glad to hear you solved it. Honestly off the top of my head I'm not sure there's any (easy) way around it. If you're interested, here's the code that's responsible for that behaviour: https://github.com/share/sharedb/blob/c711cfcb777213d193b1f4a101125e8f6e8e6864/lib/client/doc.js#L584-L620
> 
> 
> 
> You'll see it performs this check:
> 
> 
> 
> ```js
> 
> if (!source && this.type === types.defaultType && op.op.length > 1)
> 
> ```
> 
> 
> 
> and then "shatters" the op.
> 
> 
> 
> So one potential way around this would be to register second `json0` type, which _isn't_ the default type, and use that, but it's pretty hacky.
> 
> 
> 
> You could also try switching to the [`json1`][1] type, which seems to have more semantic operations (including replace).
> 
> 
> 
> The final thing — which I _think_ is what you've landed on? — is to buffer incoming ops and batch them together.
> 
> 
> 
> Of course you could always raise a pull request to disable this behaviour using some sort of config flag!
> 
> 
> 
> [1]: https://github.com/ottypes/json1

thx

**Comment by alecgibson** [comment]
@aeroyu this should now be fixed in `v1.5.0` — you can listen for the `op batch` event instead of `op`

**Comment by aeroyu** [comment]
thx ， great help to me
